UOP1_Project/Assets/PlayerSpawnSystem/PlayerLevelChange.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.SceneManagement;
+
+
+//The main "Spawn System" management class. Call the
+public class PlayerLevelChange : MonoBehaviour
+{
+    public static PlayerLevelChange main;
+    [SerializeField] List<PlayerSpawnPoint> spawnPoints = new List<PlayerSpawnPoint>(); //Contains all the spawn points in the current scene
+
+    void Awake()
+    {
+        //========== Singleton design stuff ==========
+        if (main == null)
+        {
+            DontDestroyOnLoad(gameObject);
+            main = this;
+        }
+        else if (main != this)
+        {
+            Destroy(gameObject);
+        }
+
+        //========== Initialization ==========
+        SceneManager.sceneLoaded += OnSceneLoaded;
+    }
+
+    void OnSceneLoaded(Scene scene, LoadSceneMode mode)
+    {
+        ClearSpawnPoints();
+    }
+
+    void ClearSpawnPoints()
+    {
+        spawnPoints.Clear();
+    }
+
+    public void AddSpawnPoint(PlayerSpawnPoint sp)
+    {
+        spawnPoints.Add(sp);
+    }
+
+    //Teleports/'Spawns' the player at the specified spawn point in the current scene
+    public void SpawnPlayerAtSpawnPoint(Protagonist player, string spawnPointLabel)
+    {
+        foreach(PlayerSpawnPoint sp in spawnPoints)
+        {
+            if(sp.label == spawnPointLabel)
+            {
+                Spawn(player, sp.transform.position, sp.transform.rotation);
+                return;
+            }
+        }
```

Comment:
This can be reduced to a single LINQ call, here's a suggestion

```suggestion
        PlayerSpawnPoint sp = spawnPoints.First(point => point.label == spawnPointLabel);
        Spawn(player, sp.transform.position, sp.transform.rotation);
```

Since the code is setup to only allow for a single spawn point per label you might instead want to use a `Dictionary<string,PlayerSpawnPoint>` object to store all spawnpoints which takes it down from a linear search to instant access.

If there can be multiple spawn points with the same name then this code wouldn't work but at the same time that doesn't sound like a good option so keeping it focused on just 1 name seems fine to me

Reply:
I've implemented your other suggestions, thank you for the feedback!

I figured that there would only be a couple of spawn points in the scene, that wouldn't be so large as to need a system that prevents duplicates, but perhaps I should add a warning in the console if duplicates are found, just to be safe?

Reply:
A debug warning could work, just be mindful of where you perform the check so you don't accidentally spam the console :)

Reply:
I think it's an interesting question in the appropriate times to use LINQ functions. With enough usage they create a noticeable overhead; they are also just implement loops with multiple iterators, dereferencing, delegate invokes, etc... in exchange for just somewhat cleaner code.

Reply:
Yeah they have to be used sparingly, with small parts like this it's mostly fine but once there are larger LINQ setups it might cause issues